pre_commit/languages/download.py

```diff
@@ -0,0 +1,258 @@
+            )
+
+        self._hasher = hashlib.new(self.algorithm)
+
+    @property
+    def value(self) -> str:
+        return self._value
+
+    @property
+    def algorithm(self) -> str:
+        return self._algorithm
+
+    @property
+    def checksum(self) -> str:
+        return self._checksum
+
+    @property
+    def digest(self) -> str:
+        return b64encode(self._hasher.digest()).decode('utf-8')
+
+    def update(self, data: bytes) -> None:
+        return self._hasher.update(data)
+
+    def check_content(self) -> None:
+        if self.digest != self.checksum:
+            raise ContentCorruptionError(f"""SRI checker content corruption,
+the digest is {self.algorithm}-{self.digest}""")
+
+    def __str__(self) -> str:
+        return self.value
+
+
+@dataclass(frozen=True)
+class URI:
+    value: str
+
+    def __post_init__(self) -> None:
+        url = urlparse(self.value)
+        if not all([url.scheme, url.netloc]):
+            raise ValueError(f'Invalid URI: {self.value}')
+
+    def __str__(self) -> str:
+        return self.value
+
+
+@dataclass(frozen=True)
+class Metadata:
+    value: str
+
+    @property
+    def parts(self) -> tuple[str, str, str, str]:
+        first, second, third, fourth = self.value.splitlines()
+        return (first, second, third, fourth)
+
+    @property
+    def platform(self) -> Platform:
+        platform, _, _, _ = self.parts
+        return Platform(platform)
+
+    @property
+    def sri(self) -> SRI:
+        _, sri, _, _ = self.parts
+        return SRI(sri)
+
+    @property
+    def uri(self) -> URI:
+        _, _, uri, _ = self.parts
+        return URI(uri)
+
+    @property
+    def filename(self) -> PurePath:
+        _, _, _, path = self.parts
+        return PurePath(path)
+
+
+def io_buffer(io: IO[AnyStr], size: int = 4096) -> Iterator[AnyStr]:
+    """Read chunks from a stream.
+    :param io: the stream to read chunks from
+    :param size: the size of chunks to read
+    :returns: iterator of the chunks
+    """
+    buffer = io.read(size)
+    while buffer:
+        yield buffer
+        buffer = io.read(size)
+
+
+def download(uri: URI, sri: SRI, filename: Path) -> None:
+    filename.parent.mkdir(parents=True, exist_ok=True)
+    with urlopen(str(uri)) as ws:
+        with filename.open('wb') as fp:
+            for buf in io_buffer(ws):
+                fp.write(buf)
+                sri.update(buf)
+            fp.flush()
+    sri.check_content()
+    chmod(filename, stat.S_IRUSR | stat.S_IXUSR)
+
+
+def install_environment(
+        prefix: Prefix,
+        version: str,
+        additional_dependencies: Sequence[str],
+) -> None:
+    host = Platform.host()
+    for dep in additional_dependencies:
+        m = Metadata(dep)
+        if host == m.platform:
+            envdir = Path(
+                lang_base.environment_dir(
+                    prefix, ENVIRONMENT_DIR,
+                    version,
+                ),
+            )
```

Comment:
You may want a helper function for this snippet as you have it twice in this file?

```
def environment_dir(...) -> Path:
```

Reply:
I see this happening elsewhere in the codebase, e.g. for Haskell [L28](https://github.com/pre-commit/pre-commit/blob/f56b75dd77a4b6d58cb67a8b5ad626c67837ece6/pre_commit/languages/haskell.py#L28
) and [L39](https://github.com/pre-commit/pre-commit/blob/f56b75dd77a4b6d58cb67a8b5ad626c67837ece6/pre_commit/languages/haskell.py#L39), for Lua [L48](https://github.com/pre-commit/pre-commit/blob/main/pre_commit/languages/lua.py#L48) and [L60](https://github.com/pre-commit/pre-commit/blob/main/pre_commit/languages/lua.py#L60). I don't think this is feasible, as ``prefix`` and ``version`` are all arguments from the parent function, they aren't available outside of the calling function.

Reply:
I don't understand. You would pass the argument to the function?

```py
def environment_dir(prefix: str, version: str) -> Path:
    return Path(lang_base(prefix, "download", version))
```

Then it is used here as so:

```suggestion
            envdir = environment_dir(prefix, version)
```

It can be used in the other two places in here as well.

So rather than having a `ENVIRONMENT_DIR` global, you can just have a helper function.